mantidproject/mantid
# Reflectometry GUI restructuring 1: Settings tabs

**Description of work.**
Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR includes replacement of the `Settings` tab into separate `Event` and `Instrument` tabs. The new tabs use MVP models and improved validation.

This PR has a large number of additions and deletions due to a lot of code having been moved around. In particular:
- the contents of `Presenters/` has been moved to `GUI/RunsTable/`, with the `Batch` components being renamed to `JobsTable`
- the `Event` presenters/views have been replaced by the new components in `GUI/Event`
- the `SettingsTab` components have been removed and replaced by two new components, `GUI/Experiment` and `GUI/Instrument`.
- Most of the slicing code has been removed as this is no longer required (it doesn't belong in the GUI logic so has been moved to an algorithm which we will call once we implement processing - not done yet)
- The `ReflDataProcessor...` items have been removed as the data processor widget has been replaced by the JobTreeView.

**Report to:** nobody

**To test:**

Note that this is work in progress on a feature branch so not all functionality is available. Here are some suggestions for tests:
- Open the `ISIS Reflectometry` interface.
- Try the buttons above the table to add/delete rows/groups
- Try entering values in the table and copy/paste etc.
- Try keyboard shortcuts like Enter to add a row and Ctrl-Enter to add a child row
- Try entering an investigation ID e.g. `INTER` and `1120015` and click `Search`. Note that catalog login is not currently implemented so you need to first login to the catalog via the main Catalog menu.
- The search should succeed and show all the runs for the investigation. Select some of them and click `Transfer`. They should be added to the main table and grouped by title with runs for different angles on different rows. If runs have the same title and angle they will appear on the same line as summed, e.g. `11935+11936`
- Click `Autoprocess`. This will copy all runs for the experiment into the main table and start an endless progress bar. Otherwise it is not implemented and currently does nothing. Note that Pause is not currently implemented.
- Click `Start monitor`. This should start live data monitoring for the current instrument. Once it's running `Stop monitor` should be available to stop it. Note that this does *not* currently pick up settings from the settings tabs so will just run with the algorithm defaults.
- Try editing inputs on the Event, Experiment and Instrument tabs. These values don't currently do anything because processing is not implemented, but check that e.g. invalid values are disallowed or validated appropriately
- Check that the Save tab works - load a workspace and click `Refresh` to see it in the GUI. Try changing some properties and clicking `Save`. It should be saved in the given save path.
- Try clicking `Batch 2` on the vertical tabs on the left. This gives a new, blank set of horizontal tabs where you can enter entirely different set of settings. Try editing some things and toggling back to the original batch.
- Try adding new batches from the `Batch` menu and deleting batches with the close buttons on the tabs.

** Known issues **
- There is a crash if you delete the last row in the table.
- There is a known crash bug if you press Ctrl-Enter more than once quickly
- If you copy a group then paste it onto a child row of an existing group, the tree extends to more than 2 levels deep (it should be 2 max)
- Processing and Autoprocess are not implemented
- Catalog login is not implemented
- The changes from #23549 to one of the deleted files have not been incorporated in the new files yet. I'll do this as a separate Pr.

Fixes #23043 and Refs #22263

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
